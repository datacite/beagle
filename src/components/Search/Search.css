.Search {
    margin-top: 1em;
    width: 100%;
}

.Search-input {
    margin-bottom: 2em;
}

.Search-results ul {
    list-style: none;
    padding: 0;
}

.Search-filters .badge {
    float: right;
}

.Search-filters .list-group {
    margin-top: 2em;
    margin-bottom: 2em;
}