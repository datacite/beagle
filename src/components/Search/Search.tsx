import React from "react";
import Form from "react-bootstrap/Form";
import FormControl from "react-bootstrap/FormControl";
import Button from "react-bootstrap/Button";
import {
  ServiceListingData,
  ServiceListingItem,
} from "../ServiceListingItem/ServiceListingItem";
import Error from "../Error/Error";
import { useQueryState } from "use-location-state";
import { useQuery, gql } from "@apollo/client";

import "./Search.css";
import {
  Container,
  Row,
  Col,
  Badge,
  ListGroup,
  InputGroup,
} from "react-bootstrap";

interface ServiceQueryResult {
  id: string;
  doi: string;
  totalCount: number;
  titles: [
    {
      title: string;
    }
  ];
  descriptions: [
    {
      description: string;
    }
  ];
  creators: [
    {
      name: string;
    }
  ];
}

interface PageInfo {
  endCursor: string;
  hasNextPage: boolean;
}

interface ServiceEdge {
  node: ServiceQueryResult;
}

interface Facet {
  count: number;
  id: string;
  title: string;
}

interface ServiceQueryData {
  services: {
    __typename: String;
    edges: ServiceEdge[];
    pageInfo: PageInfo;
    pidEntities: Facet[];
    fieldsOfScience: Facet[];
    totalCount: Number;
  };
}

interface ServiceQueryVar {
  query: string;
  pidEntity: string;
  fieldOfScience: string;
}

export const SERVICES_GQL = gql`
  query getServicesQuery(
    $query: String!
    $pidEntity: String
    $fieldOfScience: String
  ) {
    services(
      query: $query
      repositoryId: "datacite.services"
      pidEntity: $pidEntity
      fieldOfScience: $fieldOfScience
      first: 500
    ) {
      edges {
        node {
          id
          doi
          titles {
            title
          }
          descriptions {
            description
            descriptionType
          }
          creators {
            name
          }
        }
      }
      pageInfo {
        endCursor
        hasNextPage
      }
      pidEntities {
        id
        title
        count
      }
      fieldsOfScience {
        id
        title
        count
      }
      totalCount
    }
  }
`;

export const Search: React.FunctionComponent = () => {
  const [searchQuery, setSearchQuery] = useQueryState("search", "");
  const [internalSearchQuery, setInternalSearchQuery] = React.useState("");

  const onSearchChange = (
    e: React.ChangeEvent<FormControl & HTMLInputElement>
  ): void => {
    setInternalSearchQuery(e.currentTarget.value);
  };

  const onReset = (e: React.MouseEvent<FormControl & HTMLButtonElement>) => {
    setInternalSearchQuery("");
  };

  const handleSearch = () => {
    setSearchQuery(internalSearchQuery);
  };

  return (
    <Container className="Search">
      <Form key="search" className="Search-input" onSubmit={handleSearch}>
        <InputGroup>
          <InputGroup.Prepend>
            <Button onClick={onReset} className="bg-transparent">
              <i className="fa fa-times"></i>
            </Button>
          </InputGroup.Prepend>
          <FormControl
            onChange={onSearchChange}
            size="lg"
            type="text"
            placeholder="Search"
            value={internalSearchQuery}
          />
          <InputGroup.Append>
            <Button variant="secondary" type="submit">
              Search
            </Button>
          </InputGroup.Append>
        </InputGroup>
      </Form>

      <SearchResults searchQuery={searchQuery}></SearchResults>
    </Container>
  );
};

type Props = {
  searchQuery: string;
};

export const SearchResults: React.FunctionComponent<Props> = (props) => {
  const [pidTypes, setPidTypes] = useQueryState<string[]>("pidtypes", []);
  const [disciplines, setDisciplines] = useQueryState<string[]>(
    "disciplines",
    []
  );
  const [searchResults, setSearchResults] = React.useState<
    ServiceListingData[]
  >([]);

  const { loading, error, data, refetch } = useQuery<
    ServiceQueryData,
    ServiceQueryVar
  >(SERVICES_GQL, {
    errorPolicy: "all",
    variables: {
      query: props.searchQuery,
      pidEntity: pidTypes.toString(),
      fieldOfScience: disciplines.toString(),
    },
  });

  const toggleFilters = (filters: string[], setFunc: Function, id: string) => {
    const activeFilters = filters.filter((t) => t !== id);
    setFunc(filters.includes(id) ? activeFilters : [...filters, id]);
  };

  const resetFilters = (
    e: React.MouseEvent<FormControl & HTMLButtonElement>
  ) => {
    setDisciplines([]);
    setPidTypes([]);
  };

  React.useEffect(() => {
    refetch({
      query: props.searchQuery,
      pidEntity: pidTypes.toString(),
      fieldOfScience: disciplines.toString(),
    });

    const results: ServiceListingData[] = [];
    if (data) {
      data.services.edges.map((edge) => {
        let dataset = edge.node;
        let name = "No Title";
        if (dataset.titles.length > 0) {
          name = dataset.titles[0].title;
        }

        let description = "";
        if (dataset.descriptions.length > 0) {
          description = dataset.descriptions[0].description;
        }
        let creators = [""];
        creators = dataset.creators.map((c) => c.name);

        results.push({
          id: dataset.id,
          doi: dataset.doi,
          name: name,
          description: description,
          creators: creators,
        });

        return results;
      });

      setSearchResults(results);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.searchQuery, data, refetch]);

  const renderResults = () => {
    if (loading) return <p>Loading...</p>;

    if (error)
      return (
        <Error
          title="Something went wrong."
          message="Unable to load services."
        />
      );

    if (!data) return "";

    return (
      <div className="Search-results">
        <p>Num services: {data.services.totalCount}</p>
        <ul>
          {searchResults.map((item) => (
            <li key={item.id}>
              <ServiceListingItem service={item}></ServiceListingItem>
            </li>
          ))}
        </ul>
      </div>
    );
  };

  const renderFilters = () => {
    if (!data) return "";

    return (
      <div className="Search-filters">
        <div>
          <h4>Filters</h4>
          <Button onClick={resetFilters} variant="link" className="p-0">
            <i className="fa fa-times"> Reset</i>
          </Button>
        </div>
        {data.services.pidEntities.length > 0 && (
          <ListGroup>
            <h5>PID Types</h5>
            {data.services.pidEntities.map((item) => (
              <ListGroup.Item
                action
                key={item.id}
                active={pidTypes.includes(item.id)}
                onClick={() => toggleFilters(pidTypes, setPidTypes, item.id)}
              >
                {item.title}
                <Badge pill variant="secondary">
                  {item.count}
                </Badge>
              </ListGroup.Item>
            ))}
          </ListGroup>
        )}

        {data.services.fieldsOfScience.length > 0 && (
          <ListGroup>
            <h5>Disciplines</h5>
            {data.services.fieldsOfScience.map((item) => (
              <ListGroup.Item
                action
                key={item.id}
                active={disciplines.includes(item.id)}
                onClick={() =>
                  toggleFilters(disciplines, setDisciplines, item.id)
                }
              >
                {item.title}
                <Badge pill variant="secondary">
                  {item.count}
                </Badge>
              </ListGroup.Item>
            ))}
          </ListGroup>
        )}
      </div>
    );
  };

  return (
    <Container>
      <Row>
        <Col sm={4}>{renderFilters()}</Col>
        <Col sm={8}>{renderResults()}</Col>
      </Row>
    </Container>
  );
};

export default Search;
