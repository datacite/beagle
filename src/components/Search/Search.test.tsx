import React from 'react';
import { render, wait } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import {
  BrowserRouter as Router,
} from "react-router-dom";
import { SERVICES_GQL, Search } from './Search';

const mock =
{
  request: {
    query: SERVICES_GQL,
    variables: {
      query: '',
      pidEntity: '',
      fieldOfScience: ''
    },
  },
  result: {
    data: {
      services: {
        edges: [
          {
            node: {
              id: 'https://handle.test.datacite.org/10.24427/141E-N846',
              doi: '10.24427/141E-N846',
              titles: [{
                title: "This is a test service."
              }],
              descriptions: [{
                description: "This test service is rendered using mocked data.",
                descriptionType: ""
              }],
              creators: [{
                name: "Example Inc"
              }]
            }
          },
          {
            node: {
              id: 'https://handle.test.datacite.org/10.24427/8gt9-nk59',
              doi: '10.24427/8gt9-nk59',
              titles: [{
                title: "This is a test service."
              }],
              descriptions: [{
                description: "This test service is rendered using mocked data.",
                descriptionType: ""
              }],
              creators: [{
                name: "Acme Inc"
              }]
            },
          }
        ],
        pageInfo: {
          endCursor: "NA",
          hasNextPage: false,
        },
        pidEntities: [
          {
            id: "dataset",
            title: "Dataset",
            count: 1
          },
          {
            id: "publication",
            title: "Publication",
            count: 2
          }
        ],
        fieldsOfScience: [
          {
            id: "computer_and_information_sciences",
            title: "Computer and information sciences",
            count: 1
          },
          {
            id: "physical_sciences",
            title: "Physical sciences",
            count: 1
          }
        ],
        totalCount: 2,
      }
    },
  },
};


test('renders', async () => {
  const { container } = render(
    <MockedProvider mocks={[mock, mock]} addTypename={false}>
      <Router>
        <Search />
      </Router>
    </MockedProvider >
  );

  expect(container).toMatchSnapshot();

  // wait for content
  await wait(() => container);

  // take a snapshot of the rendered content (error or data)
  expect(container).toMatchSnapshot();
});
