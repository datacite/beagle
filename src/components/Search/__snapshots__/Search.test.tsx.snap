// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div
    class="Search container"
  >
    <form
      class="Search-input"
    >
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <button
            class="bg-transparent btn btn-primary"
            type="button"
          >
            <i
              class="fa fa-times"
            />
          </button>
        </div>
        <input
          class="form-control form-control-lg"
          placeholder="Search"
          type="text"
          value=""
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-secondary"
            type="submit"
          >
            Search
          </button>
        </div>
      </div>
    </form>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-4"
        />
        <div
          class="col-sm-8"
        >
          <p>
            Loading...
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders 2`] = `
<div>
  <div
    class="Search container"
  >
    <form
      class="Search-input"
    >
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <button
            class="bg-transparent btn btn-primary"
            type="button"
          >
            <i
              class="fa fa-times"
            />
          </button>
        </div>
        <input
          class="form-control form-control-lg"
          placeholder="Search"
          type="text"
          value=""
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-secondary"
            type="submit"
          >
            Search
          </button>
        </div>
      </div>
    </form>
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-4"
        >
          <div
            class="Search-filters"
          >
            <div>
              <h4>
                Filters
              </h4>
              <button
                class="p-0 btn btn-link"
                type="button"
              >
                <i
                  class="fa fa-times"
                >
                   Reset
                </i>
              </button>
            </div>
            <div
              class="list-group"
            >
              <h5>
                PID Types
              </h5>
              <button
                class="list-group-item list-group-item-action"
              >
                Dataset
                <span
                  class="badge badge-pill badge-secondary"
                >
                  1
                </span>
              </button>
              <button
                class="list-group-item list-group-item-action"
              >
                Publication
                <span
                  class="badge badge-pill badge-secondary"
                >
                  2
                </span>
              </button>
            </div>
            <div
              class="list-group"
            >
              <h5>
                Disciplines
              </h5>
              <button
                class="list-group-item list-group-item-action"
              >
                Computer and information sciences
                <span
                  class="badge badge-pill badge-secondary"
                >
                  1
                </span>
              </button>
              <button
                class="list-group-item list-group-item-action"
              >
                Physical sciences
                <span
                  class="badge badge-pill badge-secondary"
                >
                  1
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          class="col-sm-8"
        >
          <div
            class="Search-results"
          >
            <p>
              Num services: 
              2
            </p>
            <ul>
              <li>
                <div
                  class="card"
                >
                  <div
                    class="card-body"
                  >
                    <div
                      class="card-title h5"
                    >
                      <a
                        href="/services/10.24427/141E-N846"
                      >
                        This is a test service.
                      </a>
                    </div>
                    <div
                      class="mb-2 text-muted card-subtitle h6"
                    >
                      Example Inc
                    </div>
                    <p
                      class="card-text"
                    >
                      This test service is rendered using mocked data.
                    </p>
                    <a
                      class="card-link"
                      href="https://handle.test.datacite.org/10.24427/141E-N846"
                    >
                      Access Service
                    </a>
                  </div>
                </div>
              </li>
              <li>
                <div
                  class="card"
                >
                  <div
                    class="card-body"
                  >
                    <div
                      class="card-title h5"
                    >
                      <a
                        href="/services/10.24427/8gt9-nk59"
                      >
                        This is a test service.
                      </a>
                    </div>
                    <div
                      class="mb-2 text-muted card-subtitle h6"
                    >
                      Acme Inc
                    </div>
                    <p
                      class="card-text"
                    >
                      This test service is rendered using mocked data.
                    </p>
                    <a
                      class="card-link"
                      href="https://handle.test.datacite.org/10.24427/8gt9-nk59"
                    >
                      Access Service
                    </a>
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
